parachain/pallets/basic-channel/src/outbound/mod.rs

```diff
@@ -215,56 +246,101 @@ pub mod pallet {
+			let eth_message_bundles: Vec<Vec<u8>> = message_bundles
+				.clone()
+				.into_iter()
+				.map(|bundle| ethabi::encode(&vec![bundle.into()]))
+				.collect();
+
+			let commitment_hash = merkle_root::<<T as Config>::Hashing, Vec<Vec<u8>>, Vec<u8>>(
+				eth_message_bundles.clone(),
+			);
 
-			let commitment_hash = Self::make_commitment_hash(&bundle);
 			let digest_item =
 				AuxiliaryDigestItem::Commitment(ChannelId::Basic, commitment_hash.clone()).into();
 			<frame_system::Pallet<T>>::deposit_log(digest_item);
-			Self::deposit_event(Event::Committed { hash: commitment_hash, data: bundle });
 
-			T::WeightInfo::on_initialize(
-				messages.len() as u32,
-				Self::average_payload_size(&messages),
-			)
+			Self::deposit_event(Event::Committed {
+				hash: commitment_hash,
+				data: message_bundles.clone(),
+			});
+
+			set(commitment_hash.as_bytes(), &StoredLeaves(eth_message_bundles).encode());
+
+			T::WeightInfo::on_initialize(message_count, average_payload_size)
 		}
 
-		fn make_commitment_hash(bundle: &MessageBundleOf<T>) -> H256 {
-			let messages: Vec<Token> = bundle
-				.messages
-				.iter()
-				.map(|message| {
-					Token::Tuple(vec![
-						Token::Uint(message.id.into()),
-						Token::Address(message.target),
-						Token::Bytes(message.payload.to_vec()),
-					])
-				})
-				.collect();
-			let commitment = ethabi::encode(&vec![Token::Tuple(vec![
-				Token::Uint(bundle.source_channel_id.into()),
-				Token::Uint(bundle.nonce.into()),
-				Token::Array(messages),
-			])]);
-			<T as Config>::Hashing::hash(&commitment)
+		fn make_message_bundles(
+			message_queue: BoundedVec<EnqueuedMessageOf<T>, <T as Config>::MaxMessagesPerCommit>,
+		) -> Vec<MessageBundleOf<T>> {
+			let account_message_map: BTreeMap<
+				T::AccountId,
+				BoundedVec<Message<T::MaxMessagePayloadSize>, T::MaxMessagesPerCommit>,
+			> = message_queue.into_iter().fold(
+				BTreeMap::new(),
+				|mut messages_for_accounts, enqueued_message| {
+					let (account, message) = (enqueued_message.account, enqueued_message.message);
+
+					if let Some(messages) = messages_for_accounts.get_mut(&account) {
+						// We should be able to safely ignore the result here, since we can't have
+						// more messages for a single account than we have in the message queue
+						messages.try_push(message).unwrap();
+					} else {
+						let mut messages = BoundedVec::default();
+						// Safe to ignore the result because we just created the empty BoundedVec
+						messages.try_push(message).unwrap();
+						messages_for_accounts.insert(account, messages);
+					}
+
+					messages_for_accounts
+				},
+			);
+
+			let mut message_bundles: Vec<MessageBundleOf<T>> = Vec::new();
+			for (account, messages) in account_message_map {
+				let next_nonce = <Nonces<T>>::mutate(&account, |nonce| {
+					// TODO: is this the right addition to use? u64 is large, but this will make us
+					// reuse nonces when we hit u64::MAX
+					*nonce = nonce.saturating_add(1);
```

Comment:
The only better option I'm seeing is checked_add and bubbling the error up to `commit`. Is that necessary here?

Reply:
We can't really handle a nonce overflow since `commit` should never return errors. `commit` is called in an `on_initialize` hook which itself can never return errors either.

I think we need to just assume that nonce will never overflow. Indeed, for a user to encounter that case, they would need spend absolutely huge amounts on transaction fees to throw 2^64 messages through the bridge.

Still not an ideal situation for us though, but can't think of a better approach.

Reply:
If we made each nonce into something like a BigInt or even a [u64; n], we could resize the nonce as necessary. We'd need to set a max size to implement MaxEncodedLen for now, but we'd have an upgrade path.

Is it worth putting this in now? If not, do we have a place to record these limits?

Reply:
Added an [issue](https://linear.app/snowfork/issue/SNO-309) for nonce overflow